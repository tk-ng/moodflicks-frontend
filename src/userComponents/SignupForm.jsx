import { useFields } from "../hooks/useFields";
import { useNavigate } from "react-router";
import {
	Form,
	FormGroup,
	Label,
	Input,
	Button,
	Row,
	Col,
	Card,
	CardBody,
	Alert,
} from "reactstrap";

export const SignupForm = ({ signup }) => {
	const naviagte = useNavigate();
	const [formData, handleChange, resetForm, formErrors, setFormErrors] =
		useFields({
			username: "",
			password: "",
			firstName: "",
			lastName: "",
			email: "",
		});

	const handleSubmit = async (evt) => {
		evt.preventDefault();
		try {
			await signup(formData);
			resetForm();
			naviagte("/");
		} catch (err) {
			setFormErrors(err);
		}
	};

	return (
		<div className="SignupForm">
			<Row>
				<Col
					className="container"
					md={{
						offset: 3,
						size: 6,
					}}
					sm="12"
					lg={{
						offset: 4,
						size: 4,
					}}
				>
					<h3 className="mb-3">Sign Up</h3>
					<Card>
						<CardBody>
							<Form onSubmit={handleSubmit}>
								<FormGroup>
									<Label for="username">Username</Label>
									<Input
										id="username"
										name="username"
										value={formData.username}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="password">Password</Label>
									<Input
										id="password"
										name="password"
										type="password"
										value={formData.password}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="firstName">First Name</Label>
									<Input
										id="firstName"
										name="firstName"
										value={formData.firstName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="lastName">Last Name</Label>
									<Input
										id="lastName"
										name="lastName"
										value={formData.lastName}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								<FormGroup>
									<Label for="email">Email</Label>
									<Input
										id="email"
										name="email"
										type="email"
										value={formData.email}
										onChange={handleChange}
										required
									/>
								</FormGroup>
								{formErrors.length !== 0 && (
									<Alert color="danger">
										<p className="small mb-0">{formErrors}</p>
									</Alert>
								)}
								<div className="d-grid">
									<Button color="primary">Submit</Button>
								</div>
							</Form>
						</CardBody>
					</Card>
				</Col>
			</Row>
		</div>
	);
};
