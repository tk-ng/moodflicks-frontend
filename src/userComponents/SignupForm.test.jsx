import { SignupForm } from "./SignupForm";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import userEvent from "@testing-library/user-event";

const signup = vi.fn();

describe("Render the SignupForm component", function () {
	it("mounts without crashing", async function () {
		const { container, asFragment } = render(
			<MemoryRouter>
				<SignupForm />
			</MemoryRouter>
		);
		expect(container).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("renders the SignupForm properly", async function () {
		render(
			<MemoryRouter>
				<SignupForm />
			</MemoryRouter>
		);
		expect(screen.getByLabelText("Username")).toBeInTheDocument();
		expect(screen.getByLabelText("Password")).toBeInTheDocument();
		expect(screen.getByLabelText("First Name")).toBeInTheDocument();
		expect(screen.getByLabelText("Last Name")).toBeInTheDocument();
		expect(screen.getByLabelText("Email")).toBeInTheDocument();
		expect(
			screen.getByText("Submit", { selector: "button" })
		).toBeInTheDocument();
	});
});

describe("Submit the SignupForm", function () {
	it("submits the form data properly", async function () {
		const login = vi.fn();
		render(
			<MemoryRouter>
				<SignupForm signup={signup} />
			</MemoryRouter>
		);
		const user = userEvent.setup();
		await user.type(screen.getByLabelText("Username"), "testuser1");
		await user.type(screen.getByLabelText("Password"), "testpassword");
		await user.type(screen.getByLabelText("First Name"), "U1F");
		await user.type(screen.getByLabelText("Last Name"), "U1L");
		await user.type(screen.getByLabelText("Email"), "test@test.com");
		await user.click(screen.getByText("Submit", { selector: "button" }));
		expect(signup).toHaveBeenCalledWith({
			username: "testuser1",
			password: "testpassword",
			firstName: "U1F",
			lastName: "U1L",
			email: "test@test.com",
		});
	});
});
