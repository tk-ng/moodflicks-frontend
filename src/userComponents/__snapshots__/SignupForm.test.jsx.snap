// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Render the SignupForm component > mounts without crashing 1`] = `
<DocumentFragment>
  <div
    class="SignupForm"
  >
    <div
      class="row"
    >
      <div
        class="container col-sm-12 col-md-6 offset-md-3 col-lg-4 offset-lg-4"
      >
        <h3
          class="mb-3"
        >
          Sign Up
        </h3>
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <form
              class=""
            >
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="username"
                >
                  Username
                </label>
                <input
                  class="form-control"
                  id="username"
                  name="username"
                  required=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="password"
                >
                  Password
                </label>
                <input
                  class="form-control"
                  id="password"
                  name="password"
                  required=""
                  type="password"
                  value=""
                />
              </div>
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="firstName"
                >
                  First Name
                </label>
                <input
                  class="form-control"
                  id="firstName"
                  name="firstName"
                  required=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="lastName"
                >
                  Last Name
                </label>
                <input
                  class="form-control"
                  id="lastName"
                  name="lastName"
                  required=""
                  type="text"
                  value=""
                />
              </div>
              <div
                class="mb-3"
              >
                <label
                  class="form-label"
                  for="email"
                >
                  Email
                </label>
                <input
                  class="form-control"
                  id="email"
                  name="email"
                  required=""
                  type="email"
                  value=""
                />
              </div>
              <div
                class="d-grid"
              >
                <button
                  class="btn btn-primary"
                >
                  Submit
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
